### PR TITLE
Precondition Checks

### PR DESCRIPTION
**Pull request by orenwf**
Ref #44 #26 

I have added checks for preconditions when a `NullLiteral` dependent node is found to be in a `CastExpression`.

This type of ASTNode will be rejected by the seeding phase. If while `NullSeeder` is doing it's work, such a node is encountered by the `process` method during the `visit` call and an exception is thrown which carries with it a reference to the offending AST node.

This exception is caught by the top level `process` method and a utility method is called which finds the enclosing type-dependent expression surrounding the cast statement. The `IJavaElement` handle for that is resolved and it is added to the set `notRefactorable`.

### CODE REVIEW

File: edu.cuny.hunter.optionalrefactoring.core/.classpath [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
-	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
 	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
 	<classpathentry kind="src" path="src"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
```
Comment (khatchad) [inline edu.cuny.hunter.optionalrefactoring.core/.classpath]: Doesn't seem related to preconditions.

File: edu.cuny.hunter.optionalrefactoring.core/src/edu/cuny/hunter/optionalrefactoring/core/analysis/Preconditions.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import org.eclipse.jdt.core.dom.FieldDeclaration;
+import org.eclipse.jdt.core.dom.MethodInvocation;
+import org.eclipse.jdt.core.dom.ReturnStatement;
+import org.eclipse.jdt.core.dom.SingleVariableDeclaration;
+import org.eclipse.jdt.core.dom.SuperConstructorInvocation;
+import org.eclipse.jdt.core.dom.SuperMethodInvocation;
+import org.eclipse.jdt.core.dom.VariableDeclarationExpression;
+import org.eclipse.jdt.core.dom.VariableDeclarationStatement;
+
+public interface Preconditions {
```
Comment (khatchad) [inline edu.cuny.hunter.optionalrefactoring.core/src/edu/cuny/hunter/optionalrefactoring/core/analysis/Preconditions.java]: Not sure what this is for.

File: edu.cuny.hunter.optionalrefactoring.core/src/edu/cuny/hunter/optionalrefactoring/core/analysis/Preconditions.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		return true;
+	}
+	
+	public static boolean check(VariableDeclarationStatement node) {
+		// TODO Auto-generated method stub
+		return true;
+	}
+	
+	public static boolean check(SingleVariableDeclaration node) {
+		// TODO Auto-generated method stub
```
Comment (khatchad) [inline edu.cuny.hunter.optionalrefactoring.core/src/edu/cuny/hunter/optionalrefactoring/core/analysis/Preconditions.java]: I don't think we should have these TODO comments in master.

File: edu.cuny.hunter.optionalrefactoring.core/.classpath (hunk 1 above)
Comment (orenwf) [reply edu.cuny.hunter.optionalrefactoring.core/.classpath]: I've made a new clean branch off master. Now I've `cherry-picked` the first commit which introduced this change along with some related to Preconditions. How to separate?

Comment (orenwf) [reply edu.cuny.hunter.optionalrefactoring.core/.classpath]: I think I figured it out. `git reset --soft HEAD~1` to uncommit the last commit, then `git reset HEAD <pathname>` to unstage the single file I want to get out of that commit
then `git commit` the balance of them
`git stash` the file i extracted, change to a new branch off `master` and `git stash pop` and then stage and `commit`

Comment (orenwf) [reply edu.cuny.hunter.optionalrefactoring.core/.classpath]: And then from that state, do I `git cherry-pick` the subsequent commit ? But will it have the file I just removed ? Not sure what to do.

Comment (khatchad) [reply edu.cuny.hunter.optionalrefactoring.core/.classpath]: Yeah I think that's the way to do it.

Comment (khatchad) [reply edu.cuny.hunter.optionalrefactoring.core/.classpath]: Oh I see what you mean.

Comment (khatchad) [reply edu.cuny.hunter.optionalrefactoring.core/.classpath]: You *also* need to separate your *commits* better :).

Comment (khatchad) [reply edu.cuny.hunter.optionalrefactoring.core/.classpath]: Once you cherry-pick a subsequent commit, you'll probably just get a merge conflict, in which case you can remove that line again. I think that is the way to do it.

Comment (orenwf) [reply edu.cuny.hunter.optionalrefactoring.core/.classpath]: I think I figured it out now... #66 #67 #68
